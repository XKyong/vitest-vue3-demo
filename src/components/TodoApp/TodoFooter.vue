<template>
  <footer class="footer">
    <!-- This should be `0 items left` by default -->
    <span class="todo-count"
      ><strong data-testid="done-todos-count">{{ doneTodosCount }}</strong> item
      left</span
    >
    <!-- Remove this if you don't implement routing -->
    <ul class="filters">
      <li>
        <RouterLink to="/">All</RouterLink>
      </li>
      <li>
        <RouterLink to="/active">Active</RouterLink>
      </li>
      <li>
        <RouterLink to="/completed">Completed</RouterLink>
      </li>
    </ul>
    <!-- Hidden if no completed items are left â†“ -->
    <button
      v-if="clearBtnVisible"
      data-testid="clear-completed"
      class="clear-completed"
      @click="$emit('clear-completed')"
    >
      Clear completed
    </button>
  </footer>
</template>
<script setup lang="ts">
import { computed } from 'vue'

interface TodoItem {
  id: number
  text: string
  done: boolean
}

interface Props {
  todos: TodoItem[]
}

const props = withDefaults(defineProps<Props>(), {
  todos: () => [] as TodoItem[],
})

const emit = defineEmits(['clear-completed'])

const doneTodosCount = computed(
  () => props.todos.filter((todo) => !todo.done).length
)

const clearBtnVisible = computed(() => props.todos.some((todo) => todo.done))
</script>
