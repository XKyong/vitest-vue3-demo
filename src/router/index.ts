import { createRouter, createWebHashHistory } from 'vue-router'
import type { RouteRecordRaw } from 'vue-router'

import TodoApp from '@/components/TodoApp/index.vue'

const routes: readonly RouteRecordRaw[] = [
  {
    path: '/',
    component: TodoApp
  },
  {
    path: '/active',
    component: TodoApp
  },
  {
    path: '/completed',
    component: TodoApp
  },
]

const router = createRouter({
  history: createWebHashHistory(import.meta.env.BASE_URL),
  routes,
  linkActiveClass: 'selected'
})

export { routes }

export default router
