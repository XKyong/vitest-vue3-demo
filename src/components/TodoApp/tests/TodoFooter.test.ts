import { mount, flushPromises } from '@vue/test-utils'
import { createRouter, createWebHashHistory } from 'vue-router'
import TodoFooter from '../TodoFooter.vue'
import { routes } from '@/router'

import type { VueWrapper } from '@vue/test-utils'

const router = createRouter({
  history: createWebHashHistory(),
  routes: routes,
  linkActiveClass: 'selected'
})

describe('TodoApp.vue', () => {
  let wrapper: VueWrapper | null = null
  const todos = [
    { id: 1, text: 'eat', done: false },
    { id: 2, text: 'play', done: true },
    { id: 3, text: 'sleep', done: false }
  ]
  beforeEach(() => {
    wrapper = mount(TodoFooter, {
      props: { todos },
      global: {
        plugins: [router]
      }
    })
  })

  it('Done Todos count', () => {
    const vm = wrapper!.vm as InstanceType<typeof TodoFooter>
    const doneTodoCount = vm.todos.filter(todo => !todo.done).length
    const countEl = wrapper!.find('[data-testid="done-todos-count"]')
    expect(parseInt(countEl.text())).toBe(doneTodoCount)
  })

  it('Clear Completed Show', async () => {
    // const vm = wrapper!.vm as InstanceType<typeof TodoFooter>
    const button = wrapper!.find('[data-testid="clear-completed"')
    expect(button.exists()).toBeTruthy()

    // 清除所有任务的完成状态，判断 button 不存在
    // 此时不生效，因为 todos 是 prop，不允许直接修改！
    // vm.todos.forEach(t => (t.done = false))

    wrapper = mount(TodoFooter, {
      props: {
        todos: [
          { id: 1, text: 'eat', done: false },
          { id: 2, text: 'play', done: false },
          { id: 3, text: 'sleep', done: false }
        ]
      },
      global: {
        plugins: [router]
      }
    })

    // 因为 wrapper 重新生成，因此这里要重新使用 find 拿下 button
    expect(wrapper!.find('[data-testid="clear-completed"').exists()).toBeFalsy()
  })

  it('Clear Completed', async () => {
    const button = wrapper!.find('[data-testid="clear-completed"')
    await button.trigger('click')
    expect(wrapper!.emitted()['clear-completed']).toBeTruthy()
  })

  it('Router link activeClass', async () => {
    const links = wrapper!.findAllComponents({ name: 'RouterLink' })

    router.push('/active')
    await router.isReady()
    await flushPromises()
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link!.vm.to === '/active') {
        expect(link!.classes()).toContain('selected')
      } else {
        expect(link!.classes()).not.toContain('selected')
      }
    }

    router.push('/completed')
    await router.isReady()
    await flushPromises()
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link!.vm.to === '/completed') {
        expect(link!.classes()).toContain('selected')
      } else {
        expect(link!.classes()).not.toContain('selected')
      }
    }

    router.push('/')
    await router.isReady()
    await flushPromises()
    for (let i = 0; i < links.length; i++) {
      const link = links.at(i)
      if (link!.vm.to === '/') {
        expect(link!.classes()).toContain('selected')
      } else {
        expect(link!.classes()).not.toContain('selected')
      }
    }
  })

  it('snapshot', () => {
    expect(wrapper!.html()).toMatchSnapshot()
  })
})
