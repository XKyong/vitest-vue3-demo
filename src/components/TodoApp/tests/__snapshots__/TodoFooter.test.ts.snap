// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TodoApp.vue > snapshot 1`] = `
"<footer class="footer">
  <!-- This should be \`0 items left\` by default --><span class="todo-count"><strong data-testid="done-todos-count">2</strong> item left</span><!-- Remove this if you don't implement routing -->
  <ul class="filters">
    <li><a aria-current="page" href="#/" class="selected router-link-exact-active">All</a></li>
    <li><a href="#/active" class="">Active</a></li>
    <li><a href="#/completed" class="">Completed</a></li>
  </ul><!-- Hidden if no completed items are left â†“ --><button data-testid="clear-completed" class="clear-completed"> Clear completed </button>
</footer>"
`;
